src/goto-symex/field_sensitivity.cpp

```diff
@@ -0,0 +1,346 @@
+
+  if(expr.id()==ID_symbol &&
+     expr.get_bool(ID_C_SSA_symbol) &&
+     !write)
+  {
+    ssa_exprt ssa=to_ssa_expr(expr);
+    get_fields(ns, ssa, expr);
+  }
+  else if(!write &&
+          expr.id()==ID_member &&
+          to_member_expr(expr).struct_op().id()==ID_struct)
+  {
+    simplify(expr, ns);
+  }
+  else if(!write &&
+          expr.id()==ID_index &&
+          to_index_expr(expr).array().id()==ID_array)
+  {
+    simplify(expr, ns);
+  }
+  else if(write && expr.id()==ID_member)
+  {
+    member_exprt &member=to_member_expr(expr);
+
+    if(member.struct_op().id()==ID_symbol &&
+       member.struct_op().get_bool(ID_C_SSA_symbol) &&
+       ns.follow(member.struct_op().type()).id()==ID_struct)
+    {
+      ssa_exprt tmp=to_ssa_expr(member.struct_op());
+
+      member.struct_op()=tmp.get_original_expr();
+
+      tmp.type()=member.type();
+      tmp.set(ID_expression, member);
+      tmp.update_identifier();
+
+      expr.swap(tmp);
+    }
+  }
+  else if(write && expr.id()==ID_index)
+  {
+    index_exprt &index=to_index_expr(expr);
+    simplify(index.index(), ns);
+
+    if(index.array().id()==ID_symbol &&
+       index.array().get_bool(ID_C_SSA_symbol) &&
+       ns.follow(index.array().type()).id()==ID_array &&
+       index.index().id()==ID_constant &&
+       to_array_type(ns.follow(index.array().type())).size().id()==ID_constant &&
+       !to_array_type(ns.follow(index.array().type())).size().is_zero())
+    {
+      ssa_exprt tmp=to_ssa_expr(index.array());
+
+      index.array()=tmp.get_original_expr();
+
+      tmp.type()=index.type();
+      tmp.set(ID_expression, index);
+      tmp.update_identifier();
+
+      expr.swap(tmp);
+    }
+  }
+#endif
+}
+
+/*******************************************************************\
+
+Function: field_sensitivityt::get_fields
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+void field_sensitivityt::get_fields(
+  const namespacet &ns,
+  const ssa_exprt &ssa_expr,
+  exprt &dest) const
+{
+#if 1
+  const typet &followed_type=ns.follow(ssa_expr.type());
+
+  if(followed_type.id()==ID_struct)
+  {
+    const exprt &struct_op=ssa_expr.get_original_expr();
+
+    const struct_typet &type=to_struct_type(followed_type);
+
+    const struct_union_typet::componentst &components=
+      type.components();
+
+    dest=struct_exprt(ssa_expr.type());
+    dest.reserve_operands(components.size());
+
+    for(struct_union_typet::componentst::const_iterator
+        it=components.begin();
+        it!=components.end();
+        ++it)
+    {
+      member_exprt member(struct_op, it->get_name(), it->type());
+
+      ssa_exprt tmp=ssa_expr;
+
+      tmp.type()=member.type();
+      tmp.set(ID_expression, member);
+      tmp.update_identifier();
+
+      exprt tmp_dest;
+      get_fields(ns, tmp, tmp_dest);
+      dest.copy_to_operands(tmp_dest);
+    }
+  }
+  else if(followed_type.id()==ID_array &&
```

Comment:
Since you use this in a few places, is-constant-sized-array seems like a useful utility predicate